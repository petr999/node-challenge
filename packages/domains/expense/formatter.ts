import { Expense } from './entity';
import { ExpenseFormatted } from './types';

const publicFields = [
  'id', 'dateCreated', 'userId', 'merchantName', 'amountInCents', 'currency', 'amount',
];

export function secureTrim(expenses: ExpenseFormatted[], count: number): string {
  const expensesPublic = expenses.map((expense) => {
    const expenseNew = {};
    publicFields.forEach((expenseKey) => {
      expenseNew[expenseKey] = expense[expenseKey];
    });
    return expenseNew;
  });
  return JSON.stringify([expensesPublic, count]);
}

export function format(rawExpense: Expense): ExpenseFormatted {
  const { amountInCents, ...expense } = rawExpense;
  const amount = (amountInCents / 100).toFixed(2);
  return {
    amount, ...expense,
  };
}
