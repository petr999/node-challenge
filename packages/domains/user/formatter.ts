import { capitalize } from '../../utils';
import { User } from './types';

const publicFields = ['first_name', 'last_name', 'company_name'];

export function secureTrim(user: User): string {
  return JSON.stringify(user, publicFields);
}

export function format(rawUser): User {
  const { id, company_name, ssn } = rawUser;
  return {
    id,
    first_name: capitalize(rawUser.first_name),
    last_name: capitalize(rawUser.last_name),
    company_name,
    ssn,
  };
}
