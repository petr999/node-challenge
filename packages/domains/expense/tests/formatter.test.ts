import { format, secureTrim } from '../formatter';

describe('Expenses secureTrim', () => {
  test('secureTrim should remove fields that are not defined in the list of public fields', () => {
    return expect(secureTrim([
      {
        id: '98925892-8d0c-4cd3-bdbd-e2249e15900f',
        dateCreated: new Date('2022-02-22T01:01:01Z'),
        status: 'processed',
        userId: '98925892-8d0c-4cd3-bdbd-e2249e15900e',
        merchantName: 'Fred\'s',
        amount: '55.00',
        currency: 'DKK',
      },
      {
        id: '98925892-8d0c-4cd3-bdbd-e2249e15900d',
        dateCreated: new Date('2022-02-22T01:01:01Z'),
        status: 'processed',
        userId: '98925892-8d0c-4cd3-bdbd-e2249e15900e',
        merchantName: 'Fred\'s',
        amount: '55.01',
        currency: 'DKK',
      },

    ], 2)).toMatchSnapshot('expenses-securetrim');
  });
});

describe('Expenses format', () => {
  test('format should return an instance of users that fits the API model, based on the db raw value', () => {
    return expect(format({
      id: '98925892-8d0c-4cd3-bdbd-e2249e15900f',
      dateCreated: new Date('2022-02-22T01:01:01Z'),
      status: 'processed',
      userId: '98925892-8d0c-4cd3-bdbd-e2249e15900e',
      merchantName: 'Fred\'s',
      amountInCents: 5500,
      currency: 'DKK',
    })).toEqual({
      id: '98925892-8d0c-4cd3-bdbd-e2249e15900f',
      dateCreated: new Date('2022-02-22T01:01:01Z'),
      status: 'processed',
      userId: '98925892-8d0c-4cd3-bdbd-e2249e15900e',
      merchantName: 'Fred\'s',
      amount: '55.00',
      currency: 'DKK',
    });
  });
});
