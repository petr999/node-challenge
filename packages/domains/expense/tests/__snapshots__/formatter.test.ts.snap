// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expenses secureTrim secureTrim should remove fields that are not defined in the list of public fields: expenses-securetrim 1`] = `"[[{\\"id\\":\\"98925892-8d0c-4cd3-bdbd-e2249e15900f\\",\\"dateCreated\\":\\"2022-02-22T01:01:01.000Z\\",\\"userId\\":\\"98925892-8d0c-4cd3-bdbd-e2249e15900e\\",\\"merchantName\\":\\"Fred's\\",\\"currency\\":\\"DKK\\",\\"amount\\":\\"55.00\\"},{\\"id\\":\\"98925892-8d0c-4cd3-bdbd-e2249e15900d\\",\\"dateCreated\\":\\"2022-02-22T01:01:01.000Z\\",\\"userId\\":\\"98925892-8d0c-4cd3-bdbd-e2249e15900e\\",\\"merchantName\\":\\"Fred's\\",\\"currency\\":\\"DKK\\",\\"amount\\":\\"55.01\\"}],2]"`;
